### ISSUE TITLE
Peer review

### ISSUE THREAD

**Issue by pontushed** [hnen/NumberClassifier#1]
Repository cloned at 17.8.2022 22:05

The project seems to have all the building blocks in place. There is a GUI and means to train and evaluate the model. However, since there is no manual yet, I am not really sure what the GUI is supposed to do at this time. I got the training done, but the default paths in the `training-config.json` for the MNIST data had to be changed on my machine.

I am going to focus on the neural network implementation in this review.

The code is well written and has all the relevant Javadoc entries attached. The math seems to be correct. The code is also handling errors, which is nice! From what I understand, you are doing the training with mini-batch stochastic gradient descent.

Getting to an accuracy of 96%+ is great and shows that your neural network is indeed learning. As you have probably noticed, the last percent numbers to reach 100% accuracy is what everyone is trying to reach and it gets exponentially more difficult.
 
Choosing the learning rate hyperparameter is a tricky thing. You could perhaps look into optimiser methods, such as RMSProp or Adam, or otherwise implement a scheduling algorithm for changing the learning rate - let it be quite big at the start and then decrease it. This is a trial and error procedure and can easily lead to the loss function diverging if the learning rate is initially set too high, or on the other hand stop the loss function from converging if it is set too low.

Then there is the concept of early stopping. Ideally, you should do periodical checks against validation data (maybe 10-20% of the training data) and check that the loss function of the validation data is not actually increasing while the loss function of the training set is decreasing - this is an indicator of overfitting, which is bad! So do the training long enough, but stop when the overfitting is starting to happen.

You could also consider implementing the softmax function for the output layer. In this case you would have a vector of probabilities that could be useful for displaying the confidence of your prediction and perhaps the other next best candidates. Note however that the sum of the probabilities from the softmax function is always 1, so if you are predicting a number from your own input that was not part of the MNIST dataset, then the prediction is really a wild guess with a high "certainty"...

Keep up the good work!

**Comment by hnen** [comment]
Thank you for the excellent tips, they were really enlightening. I'm definitely aware that using the app may not be very convenient at this point (like things breaking if the working directory is not same in your env...), and I just couldn't make enough time in my schedule to write a manual yet, so thanks for pushing through :-)
